### PR TITLE
[ML] Linear child ownership and fault-injected lifecycle for Sandbox2

### PR DESCRIPTION
**Pull request by valeriy42**
## Summary

Rebuilds `CSandboxedProcessSpawner` around an explicit lifecycle state machine (`Prepared -> Launched -> IdentityCaptured -> Registered -> Monitoring -> TerminationRequested -> CleanupRequired -> Reaped/Failed`), stacked on #3184.

- Non-throwing kill-and-reap guard, armed immediately after launch and PID capture, holding its own owning reference to the sandbox handle so cleanup is order-independent during unwinding.
- Four injectable seams (pidfd acquisition, registry allocation, monitor-thread launch, sandbox completion) for deterministic fault injection.
- Explicit pidfd outcome classification: a kernel lacking pidfd support (`ENOSYS`) is the *only* case that falls back to signalling the sandboxee through its owned monitor handle (SIGKILL, identity-safe, no numeric-PID lookup); every other pidfd error is treated as a resource/identity failure and fails registration outright. Numeric `kill(pid, ...)` does not appear anywhere in this file.
- A CAS-controlled one-shot latch replacing two-boolean coordination for the timeout-vs-completion race.
- `CSandboxedProcessSpawnerLifecycleTest_Linux.cc`: fault-injection coverage for every pidfd class, allocation/resource failures, stale-generation protection, PID-reuse identity binding, the CAS race, descriptor-baseline cleanup, and destructor-latency.

## Notes

- `E_Launched`/`E_CleanupRequired` lifecycle states are declared but intentionally left unassigned (no natural single point without broader restructuring) — not silently dropped.
- The SIGKILL-only assumption for the owned-monitor termination path (no graceful-SIGTERM variant is buildable against the pinned Sandbox2 dependency version) was cross-checked against the vendored monitor source during review.

### CODE REVIEW

**Comment by valeriy42** [comment]
Devbox verification (Linux x86_64, real Sandbox2 toolchain): \`ml_test_sandbox\` filtered to \`CSandboxedProcessSpawnerLifecycleTest_Linux\` — **13/13 test cases passed, 372/372 assertions, exit code 0**.

This surfaced and fixed two real defects that macOS-only syntax checks couldn't catch:
- Two `BOOST_TEST_REQUIRE` call sites compared non-streamable types (`std::future_status`, a `std::map` iterator), which is a hard compile error for that macro (it tries to print both operands on failure). Switched to plain `BOOST_REQUIRE`.
- The forkserver warm-up used to stabilize the descriptor-count baseline was wired as a `BOOST_GLOBAL_FIXTURE`, whose constructor runs before Boost.Test finishes its own framework setup. Forking a real process that early corrupted the framework's internal state (every test case genuinely passed, but the run still reported a spurious "Incorrect setup: no test case executed" with a non-zero exit code). Moved the warm-up to a `std::call_once`-guarded call from inside the normal per-case fixture instead.

Also ran clang-format via the CI's own check image and fixed several workspace-planning-doc citations that had leaked into comments/commit messages (not resolvable outside this internal effort's own tracking) — both now clean.

Buildkite (macOS/Windows legs, full cross-platform CI) still pending.

**Review by edsavage — CHANGES_REQUESTED** [review]
Looking pretty good, I just spotted one thing that I think should be fixed prior to merge

`testDestructorDoesNotJoinAndReturnsUnderOneSecond` asserts elapsed < 1s via `steady_clock`. I think this will be subject to spurious failures on a stalled CI box. The property under test here is actually binary: Either a correct destructor returns in microseconds, or a regression that joins/blocks on the monitor (parked in `AwaitResult()` on a never-exiting child) hangs forever.

We could rework the test as a deadlock watchdog rather than a timing check: run `spawner.reset()` on its own thread (moving sole ownership of the `unique_ptr` in to avoid a race),  wait_for a largeish timeout (~30s, purely as a deadlock guard rather than a performance bound), and then assert on status being `future_status::ready` so we assert on completion,  not on elapsed time.This matches the watchdog pattern already used in `testTerminateChildFallsBackToKillWhenKernelUnsupportsPidfd`. If we do go this route we should rename the test to something like `testDestructorDoesNotBlockOnLiveChild`.

File: lib/sandbox/CSandboxedProcessSpawner_Linux.cc [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+        LOG_ERROR(<< "Cannot stat " << absPath << ": " << ::strerror(errno));
+        return false;
+    }
+
+    TStrVec fullArgs;
+    fullArgs.reserve(args.size() + 1);
+    fullArgs.push_back(processPath);
+    for (const std::string& arg : args) {
+        fullArgs.push_back(arg);
+    }
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/sandbox/CSandboxedProcessSpawner_Linux.cc:427]: Out of scope for PR D: `CSandboxedProcessSpawner::spawn()` is only exercised by the lifecycle unittest today; `lifecycle_signal_payload` never opens the path-bearing argv values. PR C already maps the validated host child root to `/run/elastic/ml-ipc` in the policy. Rewriting the four pipe arguments to sandbox-visible paths belongs with PR E controller routing / real `pytorch_inference` launch wiring, not the V8–V11 lifecycle gate in this PR.

File: lib/sandbox/unittest/CSandboxedProcessSpawnerLifecycleTest_Linux.cc [hunk 2]
```diff
@@ -0,0 +1,1109 @@
...
+
+    // Left as the default (empty) seam: the real kernel's pidfd_open() is
+    // expected to succeed (E_Acquired) on any CI host new enough to build
+    // Sandbox2 at all - this is the natural, un-forced positive control.
+    TSpawner spawner{TSpawner::TPidFdOpenFn{}, insertFn, monitorLaunch, awaitResultFn};
+    TPid childPid{0};
+    BOOST_TEST_REQUIRE(spawner.spawn(ML_SANDBOX2_LIFECYCLE_PAYLOAD,
+                                     childIpcArgs(childRoot), childPid));
+    BOOST_TEST_REQUIRE(childPid > 0);
+    BOOST_TEST_REQUIRE(capturedPidFd >= 0); // confirms the real kernel classified E_Acquired
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/sandbox/unittest/CSandboxedProcessSpawnerLifecycleTest_Linux.cc:608]: Keeping this as an un-forced positive control: `testTerminateChildUsesPidfdSignalWhenAcquiredAndChildSurvives` is the natural E_Acquired path on a kernel that actually provides pidfd. ENOSYS fallback is already covered by `testTerminateChildFallsBackToKillWhenKernelUnsupportsPidfd`. Skipping or branching here would let a CI host that lacks pidfd pass this suite while hiding the acquired-path behavior this case is meant to exercise.

File: lib/sandbox/CSandboxedProcessSpawner_Linux.cc [hunk 3]
```diff
@@ -0,0 +1,804 @@
...
+//! construct a std::thread running monitorBody and detach it, converting
+//! any std::system_error from either step into a false return instead
+//! of letting it propagate as an exception - the caller (spawn()) treats a
+//! false return the same way regardless of which step failed.
+bool defaultMonitorLaunch(std::function<void()> monitorBody) {
+    try {
+        std::thread monitor{std::move(monitorBody)};
+        monitor.detach();
+        return true;
+    } catch (const std::exception&) { return false; }
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/sandbox/CSandboxedProcessSpawner_Linux.cc:283]: Fixed in the worktree: `defaultMonitorLaunch` now treats construction failure as `false`, and after the thread is running a `detach()` failure releases the joinable `std::thread` handle (leak) and returns `true` so we neither `std::terminate` on unwind nor have `spawn()` Kill/reap while the monitor is already awaiting.

File: lib/sandbox/CSandboxedProcessSpawner_Linux.cc [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+std::uint64_t defaultRegistryInsert(CSandboxedProcessSpawner::SPidRegistry& registry,
+                                    core::CProcess::TPid pid,
+                                    CSandboxedProcessSpawner::SSandboxedChild child) {
+    std::lock_guard<std::mutex> lock(registry.s_Mutex);
+    const std::uint64_t generation{++registry.s_NextGeneration};
+    const auto existing = registry.s_Children.find(pid);
+    if (existing != registry.s_Children.end()) {
+        closePidFdIfOpen(existing->second.s_PidFd);
+        LOG_DEBUG(<< "Replacing stale registry entry for sandboxed pytorch_inference PID "
+                  << pid << " before registering generation " << generation);
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/sandbox/CSandboxedProcessSpawner_Linux.cc:367]: Fixed: `closePidFdIfOpen` now takes `int&` and sets the stored value to `-1` after a successful close, so a stale registry entry cannot retain a recycled fd number after replacement.

File: lib/sandbox/CSandboxedProcessSpawner_Linux.cc [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+                }
+            } catch (const std::exception& e) {
+                LOG_ERROR(<< "Monitor thread for sandboxed pytorch_inference PID "
+                          << sandboxPid << " failed: " << e.what());
+                eraseRegistryEntryOnMonitorFailure(registry, sandboxPid, generation);
+            } catch (...) {
+                LOG_ERROR(<< "Monitor thread for sandboxed pytorch_inference PID "
+                          << sandboxPid << " failed with a non-standard exception");
+                eraseRegistryEntryOnMonitorFailure(registry, sandboxPid, generation);
+            }
```
[the remark this answers is not in the document]
Comment (valeriy42) [reply lib/sandbox/CSandboxedProcessSpawner_Linux.cc:636]: Fixed: monitor catch paths now call `sandbox->Kill()` and `sandbox->AwaitResult()` before erasing the registry entry. `testMonitorAwaitResultThrowErasesRegistryAndDoesNotEscape` now throws before completion (no prior Kill/Await in the seam) and asserts the child is reaped via pidfd ESRCH/readable observation.

**Comment by valeriy42** [comment]
Addressed the destructor feedback in `432babb5c`: renamed to `testDestructorDoesNotBlockOnLiveChild`, run `spawner.reset()` on a worker thread with sole `unique_ptr` ownership, and assert `future_status::ready` after a 30s deadlock guard (same watchdog pattern as `testTerminateChildFallsBackToKillWhenKernelUnsupportsPidfd`) instead of `elapsed < 1s`.

Copilot lifecycle/cleanup threads are replied inline; IPC argv rewrite and ENOSYS skip left as out-of-scope / intentional positive control.

File: lib/sandbox/CSandboxedProcessSpawner_Linux.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (edsavage) [reply lib/sandbox/CSandboxedProcessSpawner_Linux.cc:427]: Understood, thanks Valeriy

**Comment by edsavage** [comment]
Thanks @valeriy42! I went looking for `432babb5c` and I don't think it made it onto the PR branch — the changes aren't in the current head.

Here's what I found:

- `432babb5c` does exist in the repo (authored 2026-09-21 12:56Z) and it does contain the destructor rework you described: `testDestructorDoesNotBlockOnLiveChild` with the `std::future_status::ready` deadlock watchdog, plus the spawner cleanup hardening (pidfd clearing, non-terminating monitor detach, Kill/reap before dropping the sandbox).
- But the current PR head `5215cb37a` still has the pre-review test `testDestructorDoesNotJoinAndReturnsUnderOneSecond` with `BOOST_CHECK(elapsed < std::chrono::seconds(1))` — so the watchdog rename/fix isn't present on the branch.
- Comparing `432babb5c` against the branch heads from the 21st (`1a22ab149`, `e79c6e55f`, `5215cb37a`) all report `diverged`, i.e. `432babb5c` was never an ancestor of `feature/sandbox2-pr-d-lifecycle`. The branch was force-pushed three times that day (12:54 -> 13:49 -> 18:30 -> `5215cb37a`), and the final rebase landed on a history without the review commit.

So it looks like `432babb5c` ended up on a separate/local branch rather than the PR branch. Could you cherry-pick it onto the current head and push?

```
git cherry-pick 432babb5c7e38cb6beb9bbe0f5d4b2d5c992a9aa
```

The diff is small (spawner `+29/-9`, test `+30/-10`), though there may be minor conflicts since the branch has been rebased since. Thanks!

**Comment by edsavage** [comment]
> Addressed the destructor feedback in `432babb5c`: renamed to `testDestructorDoesNotBlockOnLiveChild`, run `spawner.reset()` on a worker thread with sole `unique_ptr` ownership, and assert `future_status::ready` after a 30s deadlock guard (same watchdog pattern as `testTerminateChildFallsBackToKillWhenKernelUnsupportsPidfd`) instead of `elapsed < 1s`.
> 
> Copilot lifecycle/cleanup threads are replied inline; IPC argv rewrite and ENOSYS skip left as out-of-scope / intentional positive control.

Thanks @valeriy42! I went looking for `432babb5c` and I don't think it made it onto the PR branch, the changes aren't in the current head.

However,

- `432babb5c` _does_ exist in the repo (authored 2026-09-21 12:56Z) and it does contain the destructor rework you described: `testDestructorDoesNotBlockOnLiveChild` with the `std::future_status::ready` deadlock watchdog, plus the spawner cleanup hardening (pidfd clearing, non-terminating monitor detach, Kill/reap before dropping the sandbox).
- But the current PR head `5215cb37a` still has the pre-review test `testDestructorDoesNotJoinAndReturnsUnderOneSecond` with `BOOST_CHECK(elapsed < std::chrono::seconds(1))`, so the watchdog rename/fix isn't present on the branch.
- Comparing `432babb5c` against the branch heads from the 21st (`1a22ab149`, `e79c6e55f`, `5215cb37a`) all report `diverged`, i.e. `432babb5c` was never an ancestor of `feature/sandbox2-pr-d-lifecycle`. The branch was force-pushed three times that day (12:54 -> 13:49 -> 18:30 -> `5215cb37a`), and the final rebase landed on a history without the review commit, so I suspect it got clobbered somewhere along the way or it ended up on a separate/local branch rather than the PR branch. Could you cherry-pick it onto the current head and push?

**Comment by valeriy42** [comment]
Cherry-picked the review fix onto current head `5215cb37a` as `ff6a40b98` (content from `432babb5c`).

- Destructor test is now `testDestructorDoesNotBlockOnLiveChild` with the 30s `future_status::ready` deadlock watchdog (same pattern as `testTerminateChildFallsBackToKillWhenKernelUnsupportsPidfd`), not `elapsed < 1s`.
- Also includes the spawner cleanup hardening from that commit (pidfd clearing after close, non-terminating monitor detach path, Kill/reap before registry erase on monitor await failure).

@edsavage — should be on the PR branch now; thanks for catching the dropped commit after the rebase.

**Review by edsavage — APPROVED** [review]
LGTM
